### PR TITLE
Ensure no bos or eos added to completion, no eos added to prompt

### PR DESCRIPTION
**Pull request by snova-zoltanc**
## Summary
ensure no bos or eos added to completion, no eos added to prompt

## PR Checklist
- [X] My PR is less than 500 lines of code
- [X] I have added sufficient comment as docstrings in my code
- [X] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by snova-bol — DISMISSED** [review]

File: generative_data_prep/processors/article_tokenizer.py [hunk 1]
```diff
@@ -334,10 +334,31 @@ def tokenize(self, completion: str, prompt: Optional[str] = None, category_id: O
             if self.prompt_postfix:
                 prompt = prompt + self.prompt_postfix
             prompt_token_ids = self.tokenizer.encode(prompt)
+            # If there is an EOS token at the end of the prompt then remove it
+            if (
+                hasattr(self.tokenizer, "eos_token_id")
+                and prompt_token_ids[-1] == self.tokenizer.eos_token_id
+                and len(prompt_token_ids) >= 1
```
Comment (snova-jonathanl) [inline generative_data_prep/processors/article_tokenizer.py:341]: Maybe also check that the completion is non-empty? There's an (admittedly unlikely) corner case where the completion is empty, but the prompt is (intentionally) non-empty. Perhaps we can add this kind of case to the tests as well.

File: generative_data_prep/processors/article_tokenizer.py [hunk 2]
```diff
@@ -334,10 +334,31 @@ def tokenize(self, completion: str, prompt: Optional[str] = None, category_id: O
...
+                prompt_token_ids = prompt_token_ids[:-1]
             tokens += list(map(lambda x: Token(x, TokenTypeIds.PROMPT, category_id), prompt_token_ids))
 
         if completion:
             completion_token_ids = self.tokenizer.encode(completion)
+            # If there is a BOS token at the begining of the completion then remove it
+            if (
+                hasattr(self.tokenizer, "bos_token_id")
+                and completion_token_ids[0] == self.tokenizer.bos_token_id
+                and len(completion_token_ids) >= 1
```
Comment (snova-jonathanl) [inline generative_data_prep/processors/article_tokenizer.py:352]: Similarly, let's also avoid removal if the prompt is empty (as is the case in pretraining).

**Review by snova-jonathanl — COMMENTED** [review]
Some comments about edge cases

File: generative_data_prep/processors/article_tokenizer.py (hunk 1 above)
Comment (snova-zoltanc) [reply generative_data_prep/processors/article_tokenizer.py:341]: I think we should still remove, if you check a few lines below, we always add an EOS after every line. So if the tokenizer adds one in automatically, we should remove it before adding our default EOS

- Also keep in mind that by default prompt only sequences are removed in the `_remove_prompt_only_sequences` function

File: generative_data_prep/processors/article_tokenizer.py (hunk 2 above)
Comment (snova-zoltanc) [reply generative_data_prep/processors/article_tokenizer.py:352]: Good call! 🚀

File: generative_data_prep/processors/article_tokenizer.py (hunk 1 above)
Comment (snova-jonathanl) [reply generative_data_prep/processors/article_tokenizer.py:341]: Good catch, thank you!

**Review by snova-jonathanl — APPROVED** [review]
Thanks, lgtm now!
